src/Symfony/Component/Mailer/Transport/Smtp/EsmtpTransportFactory.php

```
@@ -38,6 +38,11 @@ public function create(Dsn $dsn): TransportInterface
 
         /** @var SocketStream $stream */
         $stream = $transport->getStream();
+        if ($sourceIp = filter_var($dsn->getOption('source_ip', false), \FILTER_VALIDATE_IP, \FILTER_FLAG_IPV6)) {
+            $stream->setSourceIp("[$sourceIp]");
+        } elseif ($sourceIp = filter_var($dsn->getOption('source_ip', false), \FILTER_VALIDATE_IP, \FILTER_FLAG_IPV4)) {
+            $stream->setSourceIp($sourceIp);
+        }
```

What should we do if the `source_ip` option is not a valid IPv4 or IPv6? Should we throw an exception?

---

I'd suggest removing the calls to filter_var indeed, and let php behave as it should when bindto is incorrect.
no need for the if/else: just call $stream->setSourceIp($dsn->getOption('source_ip', ''));

---

Since IPv6 address must be [wrapped inside square brackets](https://github.com/symfony/symfony/blob/7a16efefdee2d282cd6ccfd6ecb0b82ff5a06936/src/Symfony/Component/Mailer/Transport/Smtp/Stream/SocketStream.php#L116), would this be feasible? 
```
if (false !== ($sourceIp = $dsn->getOption('source_ip', false))) {
    $ipv6 = filter_var($sourceIp, \FILTER_VALIDATE_IP, \FILTER_FLAG_IPV6);
    $stream->setSourceIp($ipv6 ? "[$sourceIp]" : $sourceIp);
}
```
Or should the user take care of enclosing the source_ip in square brackets when specifying a IPv6 address?

Also, I would prefer not to call `$stream->setSourceIp(...)` if `source_ip=` is not specified in the DSN.
Even though it should [not change](https://github.com/symfony/symfony/blob/7a16efefdee2d282cd6ccfd6ecb0b82ff5a06936/src/Symfony/Component/Mailer/Transport/Smtp/Stream/SocketStream.php#L142) the outcome of the if-statement, using `''` does change the [default value](https://github.com/symfony/symfony/blob/7a16efefdee2d282cd6ccfd6ecb0b82ff5a06936/src/Symfony/Component/Mailer/Transport/Smtp/Stream/SocketStream.php#L31). Using `null` is also possible, but then I should add a question mark to the [setter](https://github.com/symfony/symfony/blob/7a16efefdee2d282cd6ccfd6ecb0b82ff5a06936/src/Symfony/Component/Mailer/Transport/Smtp/Stream/SocketStream.php#L120C5-L120C52) I think: `public function setSourceIp(?string $ip): static`.

Thanks again for your thoughts.

---

The user should take care of enclosing the source_ip in square brackets IMHO.
I'm fine not calling the method is the option isn't set of course

---

Code is updated and option documented here: https://github.com/symfony/symfony-docs/pull/20581